KronicDeth/intellij-elixir
# Convert ELIXIR_MODULEs in Small IDEs to Projects with Elixir Facets

Fixes #1378

# Test Plan
1. Remove project and module information: `rm -rf **/*.iml **/.idea projectFilesBackup*`
2. Install IntelliJ Elixir 10.4.0 in Small IDE
3. Open project root directory in Small IDE 
4. [VERIFY] `*.iml` files with `ELIXIR_MODULE` exist in root and `apps/*` directories
5. Close the project
6. Install IntelliJ Elixir 10.4.1-pre* in Small IDE
7. Open project root directory in Small IDE
8. [VERIFY] Conversion dialog asks to Convert project
9. Click Convert
10. [VERIFY] `*.iml` files with `ELIXIR_MODULE` are gone.
11. [VERIFY] root and `apps/*` directories show up as projects (bolded)
12. Expand External Libraries
13. [VERIFY] all deps show up as external libraries.

# Test Report
- [x] Test Plan run with Rubymine 2018.3.5: ✅ 
- [x] Test Plan run with WebStorm 2019.1: ❌ 
  WebStorm does not support attaching projects.  It cannot support umbrella projects, but the converting will clean up the incorrect `ELIXIR_MODULE`s still.
- [x] Test Plan run with PHPStorm 2019.1: ❌ 
   Could not reliably get `eex_test` to attach while it worked for `eex_test_umbrella` at the root and `eex_test_web` as a sibling. 

## Summary

Attaching multiple project is unreliable outside of Rubymine, so only allow it in Rubymine and raise a notification in other small IDEs telling users of projects with multiple OTP apps to use IntelliJ (preferred) or Rubymine.
 
# Changelog
## Enhancements
* Update local build to IDEA 2019.1

## Bug Fixes
* In Small IDEs, attach OTP applications as separate projects with Elixir facets instead of directly creating `ELIXIR_MODULE` modules in `DirectoryConfigurator`.  This matches the manual steps of opening the OTP apps as separate directories after opening the root of the project.
  * Mark directories when adding facet
  * Sync libraries when adding facet or attaching to root project.  Ensures External Libraries show up for all OTP applications.
  * Convert Small IDE proejct with `ELIXIR_MODULE`s:
     1. Remove `ELIXIR_MODULE` `*.iml` files
     2. Remove `ELIXIR_MODULE` entries root project's `modules.xml` file.
     3. [Lack of modules in project in `doOpenProject`](https://github.com/JetBrains/intellij-community/blob/1f49fc455d01f4b887463853db48e814f6700fb3/platform/platform-impl/src/com/intellij/platform/PlatformProjectOpenProcessor.java#L186-L188) will trigger `DirectoryProjectConfigurator`s to be [run](https://github.com/JetBrains/intellij-community/blob/1f49fc455d01f4b887463853db48e814f6700fb3/platform/platform-impl/src/com/intellij/platform/PlatformProjectOpenProcessor.java#L206), which will setup the project and run `DirectoryConfigurator` to attach the `apps` OTP apps and Elixir facets.
   * Only support automatic attaching multiple OTP apps as separate projects in Rubymine.  In IntelliJ normal support for multiple Modules in one Project still works.
